lib/codegen/src/constant_folding.rs

```diff
@@ -0,0 +1,240 @@
+//! Fold operations on constants at compile time.
+
+use cursor::{Cursor, FuncCursor};
+use ir::{self, InstBuilder};
+use timing;
+
+enum ConstImm {
+    Bool(bool),
+    I64(i64),
+    Ieee32(f32),
+    Ieee64(f64),
+}
+
+impl ConstImm {
+    fn unwrap_i64(self) -> i64 {
+        if let ConstImm::I64(imm) = self {
+            imm
+        } else {
+            panic!("self did not contain an `i64`.")
+        }
+    }
+
+    fn evaluate_truthiness(self) -> bool {
+        match self {
+            ConstImm::Bool(b) => b,
+            ConstImm::I64(imm) => imm != 0,
+            _ => panic!(
+                "Only a `ConstImm::Bool` and `ConstImm::I64` can be evaluated for \"truthiness\""
+            ),
+        }
+    }
+}
+
+/// Fold operations on constants.
+///
+/// It's important to note that this will not remove unused constants. It's
+/// assumed that the DCE pass will take care of them.
+pub fn fold_constants(func: &mut ir::Function) {
+    let _tt = timing::constant_folding();
+    let mut pos = FuncCursor::new(func);
+
+    while let Some(_ebb) = pos.next_ebb() {
+        while let Some(inst) = pos.next_inst() {
+            use ir::InstructionData::*;
+            match pos.func.dfg[inst] {
+                Binary { opcode, args } => {
+                    fold_numerical_binary(&mut pos.func.dfg, inst, opcode, args);
+                }
+                Branch {
+                    opcode: _,
+                    args: _,
+                    destination: _,
+                } => {
+                    fold_simple_branch(&mut pos.func, inst);
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+fn resolve_value_to_imm(dfg: &ir::DataFlowGraph, value: ir::Value) -> Option<ConstImm> {
+    let original = dfg.resolve_aliases(value);
+
+    let inst = dfg.value_def(original).unwrap_inst();
+
+    use ir::{InstructionData::*, Opcode::*};
+    match dfg[inst] {
+        UnaryImm {
+            opcode: Iconst,
+            imm,
+        } => Some(ConstImm::I64(imm.into())),
+        UnaryIeee32 {
+            opcode: F32const,
+            imm,
+        } => {
+            let imm_as_f32 = f32::from_bits(imm.bits()); // see https://doc.rust-lang.org/std/primitive.f32.html#method.from_bits for caveats
```

Comment:
Could you put the comment before this line. This line is a bit long.

Reply:
This got replaced by software floats.